### ISSUE TITLE
How to create a Validator for a class that is a collection?

### ISSUE THREAD

**Issue by edjo23** [Avanade/Beef#136]
Hi,

Is there a way to create a Validator for a class that is a collection? i.e. something that would iterate through the items in the collection and invoke a validator for each item in the collection. I know you can do this for a property of a class that is a collection but I can't see how to do this if the class that is being validated is a collection.

Thanks

**Comment by edjo23** [comment]
With the Beef.Core update to support Dictionary properties, we have some validators that are now no longer compiling due to a signature change to the `Collection` method that is chained to the `Property` method. It is now expecting an ICollection where as previously it was expecting an IEnumerable. The `Items` property is an IEnumerable.

```
    public class BookingScheduleCollectionValidator : Validator<BookingScheduleCollection>
    {
        /// <summary>
        /// Initializes a new instance of the <see cref="BookingScheduleValidator"/> class.
        /// </summary>
        public BookingScheduleCollectionValidator()
        {
            Property(x => x.Items).Collection(item: CollectionRuleItem.Create<BookingSchedule, IValidator<BookingSchedule>>());
        }
    }
```
The `BookingScheduleCollection`  class is a generated collection class.
```
    public partial class BookingScheduleCollection : EntityBaseCollection<BookingSchedule>
    {
    }
```

To me what we were doing doesn't look quite right anyhow, so just wondering what the recommended approach is.

**Comment by chullybun** [comment]
Hi @edjo23,

There is currently no means to validate a collection (or dictionary) as a validator, only as a property, as you describe. The behaviour you had before with the `Items` was accidental (non intended, and should continue to function). I have fixed the constraint to be `IEnumerable` again to facilitate.

To formalize a new `CollectionValidator` and `DictionaryValidator` have been added to enable. The validator code should look similar to the following:

``` csharp
    public class AddressCollectionValidator : CollectionValidator<AddressCollection, Address>
    {
        public AddressCollectionValidator()
        {
            MinCount = 1;
            MaxCount = 5;
            Item = CollectionRuleItem.Create(_addressValidator).DuplicateCheck(x => x.Street);
        }
    }
```

See `Beef.Core` NuGet version `4.1.15`.

Enjoy!
